quickwit-metastore/src/checkpoint.rs

```diff
@@ -0,0 +1,457 @@
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let string_to_string_map: BTreeMap<String, String> = BTreeMap::deserialize(deserializer)?;
+        let per_partition: BTreeMap<PartitionId, Position> = string_to_string_map
+            .into_iter()
+            .map(|(partition_id, position)| {
+                (PartitionId::from(partition_id), Position::from(position))
+            })
+            .collect();
+        Ok(Checkpoint { per_partition })
+    }
+}
+
+/// Error returned when trying to apply a checkpoint delta to a checkpoint that is not
+/// compatible. ie: the checkpoint delta starts from a point anterior to
+/// the checkpoint.
+#[derive(Error, Debug)]
+#[error("IncompatibleChkpt at partition: {partition_id:?} cur_pos:{current_position:?} delta_pos:{delta_position_from:?}")]
+pub struct IncompatibleCheckpoint {
+    /// One PartitionId for which the incompatibility has been detected.
+    pub partition_id: PartitionId,
+    /// The current position within this partition.
+    pub current_position: Position,
+    /// The origin position for the delta.
+    pub delta_position_from: Position,
+}
+
+impl Checkpoint {
+    /// Returns an iterator with the reached position for each given partition.
+    pub fn iter(&self) -> impl Iterator<Item = (PartitionId, Position)> + '_ {
+        self.per_partition
+            .iter()
+            .map(|(partition_id, position)| (partition_id.clone(), position.clone()))
+    }
+
+    fn check_compatibility(&self, delta: &CheckpointDelta) -> Result<(), IncompatibleCheckpoint> {
+        info!(delta=?delta, checkpoint=?self);
+        for (delta_partition, delta_position) in &delta.per_partition {
+            let position = if let Some(position) = self.per_partition.get(delta_partition) {
+                position
+            } else {
+                continue;
+            };
+            match position.cmp(&delta_position.from) {
+                Ordering::Equal => {}
+                Ordering::Less => {
+                    warn!(cur_pos=?position, delta_pos_from=?delta_position.from,partition=?delta_partition, "Some positions were skipped.");
+                }
+                Ordering::Greater => {
+                    return Err(IncompatibleCheckpoint {
+                        partition_id: delta_partition.clone(),
+                        current_position: position.clone(),
+                        delta_position_from: delta_position.from.clone(),
+                    });
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Try and apply a delta.
+    ///
+    /// `strict_mode` controls whether we accept a delta that does not chain perfectly with the
+    /// current position.
+    ///
+    ///    |    Checkpoint & Delta       | Outcome                     |
+    ///    |-----------------------------|-----------------------------|
+    ///    | ]?..a] ]a..c]               | Compatible                  |
```

Comment:
can we use rust range syntax here?

Reply:
I don't think there is a syntax for that kind of interval.